site/src/pages/WorkspacePage/hooks.ts

```diff
@@ -0,0 +1,66 @@
+import {
+  getWorkspaceBuildByNumber,
+  stopWorkspace,
+  startWorkspace,
+} from "api/api"
+import { delay } from "utils/delay"
+import { ProvisionerJob, WorkspaceBuild } from "api/typesGenerated"
+import { useMutation } from "@tanstack/react-query"
+
+export function waitForBuild(build: WorkspaceBuild) {
+  return new Promise((res, reject) => {
+    void (async () => {
+      let latestJobInfo: ProvisionerJob | undefined = undefined
+
+      while (latestJobInfo?.status !== "succeeded") {
+        const { job } = await getWorkspaceBuildByNumber(
+          build.workspace_owner_name,
+          build.workspace_name,
+          String(build.build_number),
+        )
+        latestJobInfo = job
+
+        if (
+          ["failed", "canceled"].some((status) =>
+            latestJobInfo?.status.includes(status),
+          )
+        ) {
+          return reject(latestJobInfo)
+        }
+
+        await delay(1000)
+      }
+
+      return res(latestJobInfo)
+    })()
+  })
+}
+
+export const useRestartWorkspace = (
+  setBuildError: (arg: Error | unknown | undefined) => void,
+  setLoading: (arg: boolean) => void,
+) => {
+  return useMutation({
+    mutationFn: stopWorkspace,
+    onMutate: () => setLoading(true),
+    onSuccess: async (data: WorkspaceBuild) => {
+      try {
+        await waitForBuild(data)
+        const startBuild = await startWorkspace({
+          workspaceId: data.workspace_id,
+          templateVersionId: data.template_version_id,
+        })
+        await waitForBuild(startBuild)
+
+        setLoading(false)
+      } catch (error) {
+        if ((error as WorkspaceBuild).status === "canceled") {
+          setLoading(false)
+          return
+        }
+        setBuildError(error)
+        setLoading(false)
+      }
+    },
+  })
+}
```

Comment:
Originally, I wanted to use an implementation closer to the following, which react query's [docs](https://tanstack.com/query/v4/docs/react/guides/mutations#promises) suggest is possible:
````
 const requestStopWorkspace = useMutation({ mutationFn: stopWorkspace })
 const requestStartWorkspace = useMutation({ mutationFn: startWorkspace })

  const restartWorkspace = async () => {
    try {
      await requestStopWorkspace.mutateAsync(workspace.id)
      await waitForStop(requestStopWorkspace.data)
      await requestStartWorkspace.mutateAsync({
        workspaceId,
        templateVersionId,
      })
    } catch (error) {
      ...   
    }
  }
````

but `mutateAsync` does not seem to be working and the queries are not treated asynchronously. @BrunoQuaresma is this something you've run into before?

Reply:
I put a comment above, I think you don't need to have each API call into a "query". You can have a single function to restart the workspace and wrap this function into a query.

Reply:
Hmmm that's a good idea. Let me try it!

Reply:
@BrunoQuaresma Woo that's a lot better! Thanks!